piptools/_compat/pip_compat.py

```diff
@@ -83,7 +84,24 @@ def parse_requirements(
     options: optparse.Values | None = None,
     constraint: bool = False,
     isolated: bool = False,
+    comes_from_stdin: bool = False,
 ) -> Iterator[InstallRequirement]:
+    # the `comes_from` data will be rewritten based on a number of conditions
+    #
+    #   None        do not rewrite
+    #   callable    programmatic rewrite
+    #   str         fixed rewrite
+    rewrite_comes_from: str | Callable[[str], str] | None = None
+
+    if comes_from_stdin:
+        # if data is coming from stdin, then `comes_from="-r -"`
+        rewrite_comes_from = "-r -"
+    else:
+        # if the input was a relative path, set the rewrite rule to rewrite
+        # absolute paths to be relative
+        if not _filename_is_abspath(filename):
+            rewrite_comes_from = _rewrite_absolute_comes_from_location
```

Comment:
Let's separate the variable initialization from the type declaration. Here's a suggestion that avoids having a nested else-if in favor of elif with a flipped guard check. Flat control flow is easier to follow than branchy.
```suggestion
    rewrite_comes_from: str | Callable[[str], str] | None

    if comes_from_stdin:
        # if data is coming from stdin, then `comes_from="-r -"`
        rewrite_comes_from = "-r -"
    elif _filename_is_abspath(filename):
        rewrite_comes_from = None
    else:
        # if the input was a relative path, set the rewrite rule to rewrite
        # absolute paths to be relative
        rewrite_comes_from = _rewrite_absolute_comes_from_location
```